staging/src/k8s.io/component-base/config/serializer/encode.go

```diff
@@ -0,0 +1,260 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package serializer
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"reflect"
+	"sort"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/runtime/serializer/json"
+	errorsutil "k8s.io/apimachinery/pkg/util/errors"
+	utilyaml "k8s.io/apimachinery/pkg/util/yaml"
+)
+
+// YAMLDocumentSeparator is the separator between YAML documents
+const YAMLDocumentSeparator = "---\n"
+
+// NewConfigSerializer creates a new implementation of the ConfigSerializer interface
+func NewConfigSerializer(scheme *runtime.Scheme, codecs *serializer.CodecFactory) ConfigSerializer {
+	// Create the strict deserializer
+	yamlStrictSerializer := json.NewYAMLStrictSerializer(json.DefaultMetaFactory, scheme, scheme)
+	// Store a map of known encoders, indexed by EncodingFormat
+	mediaTypeMap := map[EncodingFormat]runtime.Encoder{}
+	for _, serializerInfo := range codecs.SupportedMediaTypes() {
+		switch serializerInfo.MediaType {
+		case "application/json":
+			mediaTypeMap[JSON] = serializerInfo.Serializer
+			mediaTypeMap[JSONPretty] = serializerInfo.PrettySerializer
+		case "application/yaml":
+			mediaTypeMap[YAML] = serializerInfo.Serializer
+		}
+	}
+	return &configSerializer{scheme, codecs, yamlStrictSerializer, mediaTypeMap}
+}
+
+// ConfigSerializer provides encoding/decoding for a configuration file
+type ConfigSerializer interface {
+	DecodeInto([]byte, runtime.Object) error
+	DecodeManyInto(EncodedConfigs, DecodedConfigs) error
```

Comment:
I don't understand the use case for the DecodeManyInto/EncodeMany methods... those seem like the responsibility of a different level, not the serializer